### PR TITLE
GUI: frontend decouple get function list (#820)

### LINKED ISSUE #820: Allow taipyApp to display the list of methods that can be used for file load or trigger actions from Chalk'it
In JavaScript side, provide a new method for taipyApp, that could  be named `TaipyApp.getAllFunctions()`, similar to `taipyApp.getAllData()`. Its purpose is to provide Chalk'it with the name of the methods that can be triggered, or associated with a file load action, according to the module they belong to.

This will allow Chalk'it to propose to the user the list of available function for trigger action or file load.

image

Currently, this is not possible. This will allow file handling of demo <https://github.com/dinhlongviolin1/taipy-chalkit-demo> fully compatible with Chalk'it spirit.

### PR DESCRIPTION
**Pull request by dinhlongviolin1**
Resolve #820

### CODE REVIEW

File: frontend/taipy-gui/base/src/app.ts [hunk 1]
```diff
@@ -74,27 +76,32 @@ export class TaipyApp {
...
     }
 
     getAllData() {
-        return this.variableManager?.getAllData();
+        return this.variableData?.getAllData();
+    }
+
+    getFunctionList() {
+        const functionData = this.functionData?.getDataTree()[this.context];
+        return functionData ? Object.keys(functionData) : [];
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/base/src/app.ts]: return `Object.keys(functionData || {})` is shorter 😎

File: frontend/taipy-gui/base/src/dataManager.ts [hunk 2]
```diff
@@ -1,73 +1,73 @@
-export interface VariableModuleData {
-    [key: string]: VariableName;
+export interface ModuleData {
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/base/src/dataManager.ts]: could be `export type ModuleData = Record<string, VarName>`

File: frontend/taipy-gui/base/src/utils.ts [hunk 3]
```diff
@@ -61,12 +61,16 @@ const processWsMessage = (message: WsMessage, appManager: TaipyApp) => {
             const mc = (message.payload as Record<string, unknown>).data as string;
             window.localStorage.setItem("ModuleContext", mc);
             appManager.context = mc;
-        } else if (message.type === "GVS") {
-            const variableData = (message.payload as Record<string, unknown>).data as VariableModuleData;
-            if (appManager.variableManager) {
-                appManager.variableManager.init(variableData);
+        } else if (message.type === "GDT") {
+            const payload = message.payload as Record<string, unknown>;
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/base/src/utils.ts]: `Record<string, unknown>` => `Record<string, ModuleData>` ?

File: taipy/gui/_renderers/json.py [hunk 4]
```diff
@@ -32,6 +33,8 @@ def _default(o):
         return _date_to_string(o)
     if isinstance(o, Path):
         return str(o)
+    if isinstance(o, FunctionType):
```
Comment (FredLL-Avaiga) [inline taipy/gui/_renderers/json.py:36]: or `callable(o)` but it might be wider ?

File: taipy/gui/_renderers/json.py [hunk 5]
```diff
@@ -32,6 +33,8 @@ def _default(o):
...
+        return o.__name__
```
Comment (FredLL-Avaiga) [inline taipy/gui/_renderers/json.py:37]: Is it possible for `o` to be a lambda ?

**Review by FredLL-Avaiga — DISMISSED** [review]
😎

**Review by FabienLelaquais — DISMISSED** [review]
👌

File: frontend/taipy-gui/base/src/app.ts (hunk 1 above)
Comment (dinhlongviolin1) [reply frontend/taipy-gui/base/src/app.ts]: thanks!

File: frontend/taipy-gui/base/src/utils.ts (hunk 3 above)
Comment (dinhlongviolin1) [reply frontend/taipy-gui/base/src/utils.ts]: Haha nice one. tx Fred

File: taipy/gui/_renderers/json.py (hunk 5 above)
Comment (dinhlongviolin1) [reply taipy/gui/_renderers/json.py:37]: it is possible. if it is lambda I will return str(o)

File: frontend/taipy-gui/base/src/app.ts [hunk 6]
```diff
@@ -101,7 +101,7 @@ export class TaipyApp {
 
     getFunctionList() {
         const functionData = this.functionData?.getDataTree()[this.context];
-        return functionData ? Object.keys(functionData) : [];
+        return Object.keys(functionData || {});
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/base/src/app.ts:104]: Yeah well.. I didn't really like that...

**Review by FabienLelaquais — APPROVED** [review]
👍

File: frontend/taipy-gui/base/src/app.ts (hunk 6 above)
Comment (dinhlongviolin1) [reply frontend/taipy-gui/base/src/app.ts:104]: I like both ways and Fred is shorter so I chose it :D

Comment (FredLL-Avaiga) [reply frontend/taipy-gui/base/src/app.ts:104]: One could argue that my way is more resource consuming ...

**Pull request merged**